crates/rustc_codegen_spirv/src/linker/mod.rs

```diff
@@ -277,7 +277,10 @@ pub fn link(sess: &Session, mut inputs: Vec<Module>, opts: &Options) -> Result<L
                 .write_all(spirv_tools::binary::from_binary(&output.assemble()))
                 .unwrap();
         }
-        if opts.dce && opts.emit_multiple_modules {
+        // Run DCE again, even if emit_multiple_modules==false - the first DCE ran before
+        // structurization and mem2reg (for perf reasons), and mem2reg may remove references to
+        // invalid types, so we need to DCE again.
```

Comment:
Not sure if you want to remove the comment or tweak it, but DCE is really cheap, and removing dead code makes other passes faster (especially `mem2reg` IIUC).

For reference, I just checked on `mouse-shader` (built during the build script of `cargo build --release -p example-runner-wgpu`) and this PR gets:
```
time:   0.008; rss:  183MB ->  183MB (   +0MB)  link_dce
time:   0.000; rss:  183MB ->  183MB (   +0MB)  link_structurize
time:   0.297; rss:  183MB ->  183MB (   +0MB)  link_block_ordering_pass_and_mem2reg
time:   0.000; rss:  183MB ->  183MB (   +0MB)  link_sort_globals
time:   0.002; rss:  183MB ->  183MB (   +0MB)  link_dce_2
```
Whereas if I remove the first DCE it pretty much takes twice as long in `mem2reg`:
```
time:   0.596; rss:  183MB ->  183MB (   +0MB)  link_block_ordering_pass_and_mem2reg
time:   0.000; rss:  183MB ->  183MB (   +0MB)  link_sort_globals
time:   0.008; rss:  183MB ->  183MB (   +0MB)  link_dce_2
```

Reply:
thanks for the timing numbers! haven't looked at these in far too long, forgot how they looked~ :heart: